core/capabilities/remote/trigger_publisher.go

```diff
@@ -0,0 +1,192 @@
+	return &triggerPublisher{
+		config:        config,
+		capInfo:       capInfo,
+		capDonInfo:    capDonInfo,
+		workflowDONs:  workflowDONs,
+		dispatcher:    dispatcher,
+		messageCache:  NewMessageCache[registrationKey, p2ptypes.PeerID](),
+		registrations: make(map[registrationKey]commoncap.CapabilityRequest),
+		stopCh:        make(services.StopChan),
+		lggr:          lggr,
+	}
+}
+
+func (p *triggerPublisher) Start(ctx context.Context) error {
+	p.wg.Add(1)
+	go p.registrationCleanupLoop()
+	p.lggr.Info("TriggerPublisher started")
+	return nil
+}
+
+func (p *triggerPublisher) Receive(msg *types.MessageBody) {
+	sender := ToPeerID(msg.Sender)
+	if msg.Method == types.MethodRegisterTrigger {
+		req, err := pb.UnmarshalCapabilityRequest(msg.Payload)
+		if err != nil {
+			p.lggr.Errorw("failed to unmarshal capability request", "capabilityId", p.capInfo.ID, "err", err)
+			return
+		}
+		callerDon, ok := p.workflowDONs[msg.CallerDonId]
+		if !ok {
+			p.lggr.Errorw("received a message from unsupported workflow DON", "capabilityId", p.capInfo.ID, "callerDonId", msg.CallerDonId)
+			return
+		}
+		p.lggr.Debugw("received trigger registration", "capabilityId", p.capInfo.ID, "workflowId", req.Metadata.WorkflowID, "sender", sender)
+		key := registrationKey{msg.CallerDonId, req.Metadata.WorkflowID}
+		nowMs := time.Now().UnixMilli()
+		p.mu.Lock()
+		p.messageCache.Insert(key, sender, nowMs, msg.Payload)
+		// NOTE: require 2F+1 by default, introduce different strategies later (KS-76)
+		ready, payloads := p.messageCache.Ready(key, uint32(2*callerDon.F+1), nowMs-int64(p.config.RegistrationExpiryMs), false)
+		p.mu.Unlock()
+		if ready {
+			return
+		}
+		agg := NewDefaultModeAggregator(uint32(callerDon.F + 1))
+		aggregated, err := agg.Aggregate("", payloads)
+		if err != nil {
+			p.lggr.Errorw("failed to aggregate trigger registrations", "capabilityId", p.capInfo.ID, "workflowId", req.Metadata.WorkflowID, "err", err)
+			return
+		}
+		unmarshaled, err := pb.UnmarshalCapabilityRequest(aggregated)
+		if err != nil {
+			p.lggr.Errorw("failed to unmarshal request", "capabilityId", p.capInfo.ID, "err", err)
+			return
+		}
+		p.mu.Lock()
+		p.registrations[key] = unmarshaled
+		p.mu.Unlock()
+		p.lggr.Debugw("updates trigger registration", "capabilityId", p.capInfo.ID, "workflowId", req.Metadata.WorkflowID)
+	} else {
+		p.lggr.Errorw("received trigger request with unknown method", "method", msg.Method, "sender", sender)
+	}
+}
+
+func (p *triggerPublisher) registrationCleanupLoop() {
+	defer p.wg.Done()
+	ticker := time.NewTicker(time.Duration(p.config.RegistrationExpiryMs) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-p.stopCh:
+			return
+		case <-ticker.C:
+			now := time.Now().UnixMilli()
+			p.mu.RLock()
+			for key := range p.registrations {
+				callerDon := p.workflowDONs[key.callerDonId]
+				ready, _ := p.messageCache.Ready(key, uint32(2*callerDon.F+1), now-int64(p.config.RegistrationExpiryMs), false)
```

Comment:
Is there no locking needed around the messageCache (or internally, per key?) I see concurrent accesses between Receive and registrationCleanupLoop

Reply:
p.mu protects it (I can add a comment in the struct)